### ISSUE TITLE
how to reliably determine if a function parameter is an in, inout, out argument in vk.xml?

### ISSUE THREAD

**Issue by ville-h** [KhronosGroup/Vulkan-Docs#610]
How to reliably determine if a function parameter is an in, inout, out argument in vk.xml? Most of them can be detected from looking at const-ness, length and pointer-ness combined with the prefix of the function name. Then there are some, say "xcb_connection*", that is not an out argument even though they are marked pointer-to-non-const.

This kind of information is necessary if one wishes to generate API that tries to utilize return type over out arguments.

**Comment by krOoze** [comment]
Basically you can follow the same assumtions the [validitygenerator.py](https://github.com/KhronosGroup/Vulkan-Docs/blob/1.0/src/spec/validitygenerator.py) makes.

From top of my head:
- there should be no `inout` parameters in Vulkan (at least not yet)
- `returnedOnly` on a struct is strong indication it is `out`
- pointer to non-const is also `out`

The `xcb_connection_t*` is a bug I think, and I think I have it somewhere in PR.

**Comment by krOoze** [comment]
Basically you can follow the same assumtions the [validitygenerator.py](https://github.com/KhronosGroup/Vulkan-Docs/blob/1.0/src/spec/validitygenerator.py) makes. Though the script bothers itself only in finding out if it needs to generate valid usage language for the given parameter.

From top of my head:
- there should be no direct `inout` parameters in Vulkan, except `*Count`
- pointer to const is `in`
- `returnedOnly` on a struct (or struct parent) makes it `out`
- if it has `noautovalidity` then it should probably be `in`
- `out` parameters should be last in the function or struct, except `pNext`

The `xcb_connection_t*` apparently needs to be non-const because of the XCB API (see #295).
It is bit buggy. As you can see, even the VUs there are bit wrong too.
Any other exceptions you know of?

**Comment by oddhack** [comment]
Agreed with @krOoze . Maybe this specific API calls for some tagging, but as a general matter, we don't want to duplicate information that is derivable from other properties of the XML, which inevitably ends up generating inconsistencies. So if there are more examples of this it would be helpful to know.

**Comment by roderickc** [comment]
I just noticed this issue and would like to join on it as well. For me in/out/inout info is very important as I work on Vulkan support for Wine, roughly an API wrapper, but many parameters need conversion due to alignment differences between linux / win32 (e.g. see https://github.com/roderickc/wine-vulkan/commit/2964f6e4b11aad7c935a69af51dda001102ffbc4). For me I need to get the behaviour 100% right.

So far I have used returnedonly for out. Count obviously is inout. For me things now start to cause issues for vkGetPhysicalDeviceProperties2KHR, which clearly is inout even though it is marked returnedonly, so I would expected to have been 'out'.

**Comment by krOoze** [comment]
@roderickc Right. `sType` and `pNext` chain does not count as part of the structure for this.
So strictly speaking `sType` + `returnedonly` counts as `inout` structure. All of the fields should still be `out` except all the `sType`s and `pNext`s (incl. those in the chain).

The VUs are not generated properly here either (missing the "valid" word). The `returnedonly` probably still makes sense to indicate the struct does not need implicit VU entries for the rest of its fields.
